### PR TITLE
feat(chart): allow valueFrom in env config of components (#40095)

### LINKED ISSUE #40095: Chart: Allow to extend cleanup cronjob env with custom envs from secrets or config maps
### Official Helm Chart version

1.13.1 (latest released)

### Apache Airflow version

2.9.1

### Kubernetes Version

1.28

### Helm Chart configuration

Hi everyone,

I get some airflow config params provided via k8s secrets / configmaps (maybe from another IT department during provisioning), e.g. the remote base log folder (some s3 URI created via open tofu and provided as a k8s secret)

Currently it seems that I am not able to bring those into the cleanup cronjob because it is not rendering the `extraEnv` and `extraEnvFrom` global values, nor does it provide custom ones that are referenceable ([you can only specify a name: value custom env not a valueFrom](https://github.com/apache/airflow/blob/45bf7b972121828483f930acac60aa0751e2716f/chart/templates/_helpers.yaml#L1013) or use the standard ones built [here](https://github.com/apache/airflow/blob/1a613030e669ec8e8b0be893038da3a3ca1de9c9/chart/templates/_helpers.yaml#L54))

As the docs state I can circumvent this by creating a custom helm chart (we actually do anyways) and then adding the cronjob I need (will also do this for now), but I think this can already be native airflow helm chart functionality

[This is the job](https://github.com/apache/airflow/blob/helm-chart/1.13.1/chart/templates/cleanup/cleanup-cronjob.yaml) I am talking about

### Docker Image customizations

Yes but I think this is unrelated to this issue

### What happened

I tried to add the following env config (before checking the code actually)

```
cleanup:
  enabled: true
  env:
      - name: AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER
        valueFrom:
          secretKeyRef:
            name: secret-name
            key: AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER
      - name: AIRFLOW__DATABASE__SQL_ALCHEMY_CONN
        valueFrom:
          secretKeyRef:
            name: secret-name            key: AIRFLOW__DATABASE__SQL_ALCHEMY_CONN
```

and as this is a bad format for now it crashed with the following:

```
helm.go:84: [debug] values don't meet the specifications of the schema(s) in the following chart(s):
airflow:
- cleanup.env.0: value is required
- cleanup.env.0: Additional property valueFrom is not allowed
- cleanup.env.1: value is required
- cleanup.env.1: Additional property valueFrom is not allowed
```

Which is clear when you look up how custom_container_env is built so no surprise here

### What you think should happen instead

_No response_

### How to reproduce

Try to add some custom env and reference an already existing secret or configmap.

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dada-engineer**
This fixes #40095 and allows the users of the official helm chart to add a valueFrom specification (for secret and configmap references) to the env config of components (most of them use the `container_extra_env` functions defined in _helpers.yaml).

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
Could you please add a unit test for it (`helm_tests` dir)

**Comment by dada-engineer** [comment]
Oh sure sorry totally missed that one 🙈

File: chart/templates/_helpers.yaml [hunk 1]
```diff
@@ -1015,7 +1015,20 @@ capabilities:
...
+        name: {{ $config.valueFrom.secretKeyRef.name }}
+        key: {{ $config.valueFrom.secretKeyRef.key }}
+      {{- else if $config.valueFrom.configMapKeyRef }}
+      configMapKeyRef:
+        name: {{ $config.valueFrom.configMapKeyRef.name }}
+        key: {{ $config.valueFrom.configMapKeyRef.key }}
+      {{- end }}
+    {{- end }}
     {{- if or (eq $.Values.executor "KubernetesExecutor") (eq $.Values.executor "LocalKubernetesExecutor") (eq $.Values.executor "CeleryKubernetesExecutor") }}
   - name: AIRFLOW__KUBERNETES_ENVIRONMENT_VARIABLES__{{ $config.name }}
```
Comment (jedcunningham) [inline chart/templates/_helpers.yaml:1043]: You also need to do apply it here.

Comment (dada-engineer) [reply chart/templates/_helpers.yaml:1043]: ah nice thanks for the hint ❤️

**Comment by dada-engineer** [comment]
> Could you please add a unit test for it (`helm_tests` dir)

done

> You also need to do apply it here.

done

thanks @potiuk and @jedcunningham and sorry for the delay

**Review by romsharon98 — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
Thanks @dada-engineer! Congrats on your first commit 🎉

**Pull request merged**